### DISCUSSION TITLE
Challenge 9 - forge questions (SPOILERS!!!!)

### DISCUSSION THREAD

**Discussion by RichardPalladino** [Cyfrin/foundry-full-course-cu#385] [Q&A]
I'm attempting to use a forge script to solve the challenge, but it keeps failing.

For example, I have an interface to the challenge contract and the following function to attempt to solve the challenge:

```
    function solveChallenge(address _challengeAddress) public {
        console.log("Attempting to solve challenge on blockchain # %s", block.chainid);

            console.log("Current address: %s", msg.sender);
            console.log("Challenge address: %s", _challengeAddress);
            uint256 guess = uint256(keccak256(abi.encodePacked(msg.sender, block.prevrandao, block.timestamp))) % 100000;
            console.log("Guess: %s", guess);
            Challenge challenge = Challenge(_challengeAddress);
            challenge.solveChallenge(guess, "0xPallad1n0");
    }
```

I can see that msg.sender in the script itself is correct, but do scripts work like a regular deployed contract where the msg.sender becomes tx.origin when calling the challenge contract?

If so, are there any Foundry/forge cheatcodes or techniques so my actual wallet address is set as msg.sender when solving the challenge?

**Comment by Jaseempk** [comment]
Hey, have you tried using ``vm.prank()`` function?
 I think it can specify the address you specify as  `msg.sender`

**Reply by RichardPalladino** [reply]
Yes.  Tried that.  It gave me an error when I placed it within a `vm.startBroadcast()` stanza.

**Comment by RichardPalladino** [comment]
To be a little more clear:

A single private key may use many active public addresses that it signs for.  How do I run operations as one public address or another within a script, while operating and broadcasting to a live public chain (i.e., not a fork, though I'd also want to know how to do this on a fork).

**Comment by RichardPalladino** [comment]
Am I wrong in observing that there has been no successful challenge 9 solution?

I look on Sepolia etherscan and arbiscan and it appears all transactions are failed / reverted.

**Reply by asobiesk** [reply]
Transactions are only interactions with the human-managed accounts, not smart contracts. So, no successful transactions in etherscan means that no one has solved the challenge using a transaction from their account. People could still solve it using a smart contract and it will not be visible there since those are calls, not transactions.

**Reply by RichardPalladino** [reply]
Cool.  I reviewed the twitter bot though and it seems like no one has done it yet.  (At least not since the bot was deployed.)

**Reply by john-lawniczak** [reply]
Richard, do you have a link to the twitter bot?

Thanks

**Reply by 0xSimeon** [reply]
IIRC, @PatrickAlphaC posted that someone has solved all 15 challenges. That was like a week after the course went live.

**Reply by john-lawniczak** [reply]
You can see all [token holders](https://sepolia.etherscan.io/token/0x76b50696b8effca6ee6da7f6471110f334536321#balances). However, this contract is particular in that he said it can be solved without deploying a new contract. This is for Sepolia obviously, and not Arbitrum.

**Comment by asobiesk** [comment]
I think what you are trying to do is to force your contract to solve the challenge pretending to be your account, so that the reward NFT will be sent to your account. You can't do that. The msg.sender will always be the address of your contract calling the solveChallenge() function.

There is however a way to solve the challenge using a contract and still receive the NFT to your specified account. Here's a hint: If your contract will indisputably be the msg.sender in solveChallenge() function, it will be the one to receive NFT. Can you then maybe make your contract transfer the token to your account upon receiving it...?

**Reply by RichardPalladino** [reply]
Of course.  And that's one way to do it.  But Patrick said there was another way and I'm trying to figure out how to use Foundry's tools to do it without deploying a contract.

This question isn't just to solve the problem.  I'm trying to learn the tools so I can get better at debugging deployed/live contracts and also so I can work on smart contract security better.

**Reply by asobiesk** [reply]
Got you. I think Patrick mentioned it's possible to solve the challenge with cast. I don't know how to do it though. I think that if anyone would have solved it with cast already, it would appear as a transaction on Etherscan. But if I'm wrong and there is someone here who figured that out, please share some hints with us, I'm really curious how to do it as well.

**Reply by Ynyesto** [reply]
But why can't you do that? Can't it be done with vm.startBroadcast(vm.envUint("PRIVATE_KEY")), as seen in lesson 9?? I don't get it 😑

**Comment by shrutic11** [comment]
Try using the following:
vm.prank(your_address);
vm.startPrank(your_address) to start sending transactions as the new user and by vm.stopPrank(); to stop sending those transactions.
These should not necessarily be used within the broadcast, rather before calling the broadcast. 
vm.startBroadcast(); / vm.stopBroadcast (); should wrap the contract deployment and payable transactions. 

Please share more details on the error to help you better.

Check the following links as well:
https://book.getfoundry.sh/cheatcodes/broadcast?highlight=broadcast#broadcast
https://book.getfoundry.sh/cheatcodes/prank?highlight=prank#prank

**Reply by RichardPalladino** [reply]
I'll try that in a bit.  Nice thought.

**Comment by john-lawniczak** [comment]
I've been looking at this contract for some time to try and solve it without deploying an attack contract. I believe it (obviously) has to do with abi-decoding "correctAnswer" and the opcodes.
```
uint256 correctAnswer =
            uint256(keccak256(abi.encodePacked(msg.sender, block.prevrandao, block.timestamp))) % 100000;
```
We have the contract ABI, the deployed bytecode, and the [block info](https://sepolia.etherscan.io/block/3623753). We know the msg.sender and the block.timestamp, so we need the block.prevrandao. I've been looking at the [block.prevrandao](https://soliditydeveloper.com/prevrandao) EIP-4399 opcode; there seems to be something about the most recent block or parent block and finding the random number that way.

As of [EVM < Paris](https://docs.soliditylang.org/en/latest/cheatsheet.html#block-and-transaction-properties) **block.prevrandao** has deprecated ; it's now **block.difficulty (uint)**
``` 
block.prevrandao (uint): random number provided by the beacon chain (EVM >= Paris) (see [EIP-4399]
```
Revisiting: Using only Foundry, I think it's just a long intricate `cast call` which would be several lines long, including abi.encodepacked and a reference to the erc721 interface.

**Reply by RichardPalladino** [reply]
I don't think it has to do with `correctAnswer` directly.  We have to provide the value of `correctAnswer`, which is `uint256(keccak256(abi.encodePacked(msg.sender, block.prevrandao, block.timestamp))) % 100000`.

How do we calculate the keccak256 value of the  abi.encodePacked output msg.sender, block.prevrandao, and block.timestamp?  Each of those change depending on time and msg.sender making the call.  I haven't figured out how to pipe all those base factors into abi.encodePacked and then pipe that into keccak256 to make a cast call.

**Reply by john-lawniczak** [reply]
Yeah it's definitely tricky. 

We could also try comparing the hash outputs and see if they're similar. Patrick uses a variation of asserting strings in [Lesson 11 here](https://youtu.be/sas02qSFZ74?list=PL4Rj_WH6yLgWe7TxankiqkrkVKXIwOP42&t=30003).

I haven't spent much more time on the pure Foundry implementation though.

**Comment by 0xshivay** [comment]
Although I have not solved it, but I have come close to solving it without deploying the contract.

We can use `$(command)` in linux terminal to insert the output of one command into another command. So we can do something like this 

```
cast to-hex $(cast block latest --field timestamp --rpc-url $SEPOLIA_RPC_URL)
```

This command will first query and output the latest block's timestamp and then convert it into hex.

We can make similar commands for the rest of the `uint256(keccak256(abi.encodePacked(msg.sender, block.prevrandao, block.timestamp))) % 100000` to input the final `uint256` in contract with `cast send`.

I have also made the entire command to do so. But I am stuck at the final step. When I try to call the contract locally using the command line via `cast call` , it is executing correctly. But, once I call it via `cast send` with same command, it is reverting. Can someone tell why this is happening?

**Reply by Ynyesto** [reply]
Interesting!! I have no idea  why it may not be working but props for getting this far! Keep us posted! :-)

**Comment by tbrannt** [comment]
I just solved this challenge with `cast send`. It was pretty hard. Didn't expect that given all challenges before (1-8) were pretty simple. So if someone has specific questions, I can give some hints

**Reply by Ynyesto** [reply]
I'm sure @0x0bserve will be interested, since he already got quite far himself! https://github.com/Cyfrin/foundry-full-course-f23/discussions/385#discussioncomment-7628342

**Reply by tbrannt** [reply]
I can tell @0x0bserve why it's not working. He's retrieving the values for block n, and when using "cast call" the call will be simulated for exactly that block n. But when using "cast send" the block the tx will be included in will not be bock n but n+1 or even n+2, ...

**Reply by Ynyesto** [reply]
So did you use a different approach from his or just realise you had to use n+1?

**Reply by 0xshivay** [reply]
We have to use a different approach to solve this. My earlier approach won't work.

**Reply by jtordgeman** [reply]
@tbrannt if by any chance you are still around i'm basically stuck at the same place @0xshivay was.. i have all the calculation done but when i call the final cast call it will indeed execute in a whole different block (sometimes n+1, sometimes n+5 etc). Is there some magic way to do these calcs on the same block or maybe trigger the whole calc differently?

**Reply by tbrannt** [reply]
@jtordgeman I used a combination of tools. I wrote a small java tool calling a beacon-chain API (I used a free API from quicknode.com) to get the infos of the last block and then calculated the prevrandao value for the next block from this.
Then you have a few seconds to quickly submit your calculated answer that's based on that value. I used remix with an injected metamask provider for this and paid a high tip so my tx gets included next

**Reply by EngrPips** [reply]
Interesting.

**Reply by jtordgeman** [reply]
ah! so i think my next step is figuring out how to calculate the prevrandao value.. with my Bash script im able to calculate the answer and all on the current latest block, but when i try to submit it using `cast` it always get handled by the next block...

**Reply by tbrannt** [reply]
yeah exactly, you need to calculate the randao (not prevrandao) of the current block, which will be the prevrandao of the next block

**Reply by jtordgeman** [reply]
Just updating that im still trying to figure out how to calculate the randao :) im assuming one of the infos needed from the beacon API is the last block `randao_reveal` maybe? I cant seems to find any good documentation on the randao to be honest..

**Reply by tbrannt** [reply]
yes, `randao_reveal` goes into the calculation :) And you're right it is hard to find info on how exacatly it is done. After searching for a while I eventually found what I need in Ben Edgington's eth2 book: https://eth2book.info/

**Reply by EngrPips** [reply]
Interesting.

**Reply by ALtera21** [reply]
btw @tbrannt, do we really need the java tool to solve this challenge? do you happened to have a more newbie friendly solution?

**Reply by EngrPips** [reply]
Good question.

**Reply by tbrannt** [reply]
it doesn't matter what programming language you use to write the tool. But doing it by hand will be too slow because you only have a few seconds to calculate the next prevrandao and broadcast the transaction.

As others have mentioned there are other approaches to solve this challenge where you do not have to be that fast and won't have to write any tool

**Reply by EngrPips** [reply]
Interesting. Hey @tbrannt  how are you able to figure out what block exactly your transaction will be executed so as to use that block number in your prevrando calculation?

**Reply by tbrannt** [reply]
I'll get the data from the latest block and try to include the transaction in the very next block after that. That's why I needed the tool because it needs to be fast. 
For a high chance to get into the next block include a sufficient fee. It's well doable on sepolia but I needed a handful of runs until it worked

**Reply by EngrPips** [reply]
Basically, you are using the block number of the current block in calculating the prevrando so as to use it in the next block and the transaction execution must not exceed the next block.

**Reply by ICREE8** [reply]
Friday 10:34 pm and yes, I am also going down this rabbit hole - don't mind me guys

**Reply by jtordgeman** [reply]
If its any consolation @ICREE8 im still down this rabbit hole.. havent figured out how to calculate that damn prevrando just yet...

**Reply by EngrPips** [reply]
Rabbit holes can be infinitely deep at times.
